### DISCUSSION TITLE
usage-based-pricing

### DISCUSSION THREAD

**Discussion by giscus** [mrocklin/mrocklin.github.io#20] [Giscus]
# usage-based-pricing

We’re moving away from usage based pricing. This post talks about the pros and cons of usage based pricing in our situation, and then talks through alternatives. Usage based pricing meme Where we a...

https://matthewrocklin.com/usage-based-pricing.html

**Comment by andersy005** [comment]
@mrocklin, thanks for the thorough post! It’s been a blast working with coiled, and i'm excited to keep going. 

Could you clarify when the new pricing model goes into effect? Also, will existing teams be transitioned automatically, or will we need to handle the switch ourselves? I'm asking because I'm currently on a team with more than 3 users and the new "Essentials" tier appears to be for teams with up to 3 seats.

**Reply by mrocklin** [reply]
Good question @andersy005 !  There will be a transition period and existing customers will have a smooth grandfathering/transition experience.  Coiled platform folks should be reaching out in the next few days.

**Comment by alimanfoo** [comment]
Hi Matt,

Thanks so much for discussing this openly, very refreshing and interesting.

Our team currently runs a Nebari deployment on GCP which has worked well. Getting that up and running initially took about a week of my time, and ongoing maintenance is fairly minimal. However, I have been considering migrating to Vertex AI Workbench + Coiled as an alternative. The main attraction is that there would be zero maintenance overhead. Also it would be more flexible, i.e., easier to access a broader range of different compute resources. So I'm looking at Coiled pricing from a perspective where we already regularly use dask in the cloud and we have already got our infrastructure and workloads reasonably well-tuned.

The 10,000 hours free usage is very generous, I appreciated that. I would also be happy to pay for usage beyond that. But I find that hard to justify when the markup is more than double the actual cost of the underlying compute resources, because we always use spot instances. In your article your response to this objection doesn't really apply to us, because we are not blocked on or fighting our current infrastructure and not spending much time on maintenance and not going to accelerate our workloads by migrating to coiled (although some general dask optimisations are still possible I'm sure and would benefit us whichever infrastructure we ran on). Coiled as a basic computational service would just be more convenient and flexible.

Also usage-based pricing which is solely calculated from number of worker hours felt a bit odd initially, because I couldn't see why adding more workers to a cluster increases your running costs and so couldn't see why I should be paying more. But maybe my assumptions about your costs are wrong. My perspective also changed a bit when I thought of using adaptive scaling, that at least ensures usage costs are minimised and we are only paying markup for workers that are actually doing work. The only alternative I could think of was usage-based pricing calculated per cluster rather than per worker. But perhaps that isn't any better.

Pricing based on seats I'd find hard to justify when some team members would use the service very heavily and others only occasionally.

2c

**Reply by mrocklin** [reply]
Hi Alistair!  Good to hear from you!

Can you say more about why you want to switch?

In your response you mentioned the following:

- main attraction is that there would be zero maintenance overhead
- Also it would be more flexible, i.e., easier to access a broader range of different compute resources

Can you say more about why you care about these things?  It sounds like your current system is pretty great (near optimal efficiency, not difficult to set up, colleagues don't have an issue using it) why spend the effort to think about rearchitecting things?

**Reply by alimanfoo** [reply]
Hi Matt,

Sure, the main driver is that although Nebari doesn't need much effort to configure, deploy or maintain, it does need some. My team and I are stretched and any time we can gain back is valuable. 

Also, with Nebari, if we'd like to add a new type of VM or amount of resources allocated to notebook servers or Dask workers or schedulers, requires config change and redeployment. On top of that, there is sometimes some trial and error needed to figure out how much of a VM's resources k8s will actually allow you to allocate. With Vertex Workbench we can immediately choose from any available GCP VM type for notebooks, and same with Coiled for Dask workers and scheduler.

Hope that's useful.

**Reply by mrocklin** [reply]
Right, so the question then becomes "how much is your time maintaining the system worth?" and "what new kinds of work can you get done with the added flexibility, and how much is that worth?"

Broadly, how valuable are those things to you?  And then my question becomes "how do I construct a pricing scheme that gets somewhat close to that value".   Any thoughts on either of those two questions?

**Reply by alimanfoo** [reply]
I work in a non-profit research organisation where budgets are tight and set ahead of time for 3-5 years linked to grant funding. Also budget lines for personnel are separate from computing and so it's not easy to move "value" between the two categories. Most of the time we will simply be thinking about how to keep our cloud compute costs down so we don't blow our grant budget.

Because we have a strong focus on minimising costs, being able to do things like make use of spot instances and have clusters with adaptive scaling is all great because it helps us keep cloud compute costs low with very little wastage. E.g., we could run 50,000 hours of compute in a month and it will cost us about $500. If we ran that via Coiled it would cost about $2,500. So switching to Coiled would multiply our compute costs by a factor of 5. You might argue that $2,500 is still not very much money, but adding $2000 a month to our bill would put a significant dent in our computing budget.

In a non-profit context, the way I look at it is, I'd like to pay for Coiled (put costs on a grant), and I'd like to pay a fair price, meaning I'm not putting you out of pocket. Is it possible to define a price which means you are at the very least recouping your running costs for providing a basic Coiled computing service, without any support or consulting services? Or maybe that's also hard to do.

This is very different I'm sure from a commercial environment where someone is using Coiled to create a data product or service which they then sell. Maybe we should just be talking about an academic/non-profit pricing model, rather than trying to find a general pricing model?

**Reply by alimanfoo** [reply]
Opinionated take: Consider offering a basic Coiled computing service which is really cheap (<$0.01 per worker hour) so it becomes an almost negligible/unnoticeable item in someone's overall cloud budget. Then it becomes like water, we all get used to having it on tap and you get high adoption and usage, using Coiled is a no-brainer. If that's not viable, consider introducing tiers so price increases gradually, rather than jumping from free to a fixed value. 

If you also help some customers migrate to cloud or migrate to Dask or otherwise make big improvements in efficiency, consider selling that as a professional service, or continue to do it pro bono as a way to get customers to use the service more heavily on a long-term basis and bring their friends. But don't bake it into your pricing model for the basic computing service, because some customers don't need it.

**Reply by mrocklin** [reply]
> Consider offering a basic Coiled computing service which is really cheap (<$0.01 per worker hour) so it becomes an almost negligible/unnoticeable item in someone's overall cloud budget. Then it becomes like water, we all get used to having it on tap and you get high adoption and usage, using Coiled is a no-brainer

Yeah, I'm totally aligned that I'd like this outcome.  Lots and lots of low-cost users.  However, I don't know how to make it happen.  I don't think that setting cost very low will suffice to make enough people aware of the product.

> You might argue that $2,500 is still not very much money, but adding $2000 a month to our bill would put a significant dent in our computing budget.

Yeah, a $25k computing budget for a research group seems pretty small to me, especially if they're able to get a lot done.  I think I'm less anchored on the cloud costs than you are.  You still pay money for a car even if petrol costs decrease.

> I'd like to pay for Coiled (put costs on a grant), and I'd like to pay a fair price.  Is it possible to define a price which means you are at the very least recouping your running costs for providing a basic Coiled computing service, without any support or consulting services? Or maybe that's also hard to do.

The cost to us to provide Coiled to a new user is roughly $0.  However the price to build and run the whole organization is millions a year.  What would fair be in this context?

**Reply by mrocklin** [reply]
Constructively, I appreciate that you're in a different situation with occasional grant funding, money for compute being separate for money for people, etc..  I think that this just pushes us to think about different pricing mechanisms that work well for academic situations.  

This doesn't have to mean very low cost, instead it probably means different buying cycles (perhaps aligned with grants) or that we go to another office in the institution that buys productivity software rather than being part of a compute budget.  There are likely many ways to make something like this work if we think creatively about it.

**Reply by alimanfoo** [reply]
> The cost to us to provide Coiled to a new user is roughly $0. However the price to build and run the whole organization is millions a year. What would fair be in this context?

What if we are not a new user but a regular user running 50,000 hours compute every month through coiled. And what if we are one of N customers doing roughly the same. Could you quantify the cost of providing a basic compute service to this set of customers?

Not sure how helpful I'm being, as you say it's tricky. Just would like coiled to be a little cheaper, because the value of switching for us is much more marginal. And in my head "fair" means price is linked to cost rather than value, at least for a basic compute service with minimal support.

And you should be getting paid Megabucks for the amazing support and expertise you provide to other customers who you help to make order of magnitude efficiency or productivity gains. Distributed computing is still a black art and you guys are masters of the craft. 

Can you segment your customers better, then segment your product/service better to match?

**Reply by mrocklin** [reply]
> What if we are not a new user but a regular user running 50,000 hours compute every month through coiled. And what if we are one of N customers doing roughly the same. Could you quantify the cost of providing a basic compute service to this set of customers?

Well, let's say we spend $5mm a year to keep things running (this isn't quite right, but it's in the ballpark).  Then maybe you're thinking $5mm / N per customer?  (50,000 hours is roughly typical for a paying customer).  So if we had 100 customers then you're talking about $50k per customer.  If we had 1000 customers then you're talking about $5k per customer.

We're much closer to the 100 customers number, so if I were to sell you the product at $25k we'd still be operating at a loss (but we'd be excited to have you as a customer even at this rate!)

> And in my head "fair" means price is linked to cost rather than value

Yeah, unfortunately I'm not sure the world agrees with this 🙂 

> Can you segment your customers better, then segment your product/service better to match?

We try.  In fact our new pricing scheme is intended to do this.  A customer who doesn't need any support and is mostly using automated systems just pays the $0.05 / CPU-hour.  Customers who have teams who need more support tend to also pay a per-seat fee.  Those users who use relatively little (10,000 CPU hours or less) and need no support get to use the service entirely for free.

> because the value of switching for us is much more marginal

Yeah, and to be clear, it may not make sense for you to switch.  You all may have the expertise to maintain your own system without spending much time.  For most people the personnel costs to maintain such a system outweigh our costs, to say nothing of efficiency gains of the people using it.  You all may be really good at this though to the point where the $25k price tag doesn't outweigh your savings.

**Reply by mrocklin** [reply]
Broadly, the relevant costs on both sides aren't cloud costs.  They're human costs.  Mostly my objective in this conversation is to pivot your thinking from "I'm renting a machine on AWS" to "I and my researchers get to focus on doing science rather than this other crap" and in doing so shift from thinking about CPU-hours and towards percentage savings of an FTE.  

We employ a few FTEs so that many institutions can save fractions of FTEs.  We hope to save you a significant fraction of an FTE and we'd love to have you subsidize a small fraction of one of our FTEs as payment.

**Reply by alimanfoo** [reply]
Thanks Matt, appreciate the insights and the openness.
